usaspending_api/database_scripts/job_archive/backfill_business_categories.py

```diff
@@ -0,0 +1,288 @@
+    logging.info("Updated SQL functions in {}".format(t))
+
+
+def id_ranges(min_id, max_id):
+    for n in range(min_id, max_id + 1, CHUNK_SIZE + 1):
+        yield n, min(n + CHUNK_SIZE, max_id)
+
+
+async def get_min_max_ids(pool):
+    with Timer() as t:
+        sql = "select min(id), max(id) from transaction_normalized"
+        async with pool.acquire() as connection:
+            min_id, max_id = await connection.fetchrow(sql)
+    logging.info("Found min transaction id = {} and max transaction id = {} in {}".format(min_id, max_id, t))
+    return min_id, max_id
+
+
+async def create_chunk_queue(min_id, max_id, loop):
+    queue = asyncio.Queue(loop=loop)
+    for min_id, max_id in id_ranges(min_id, max_id):
+        await queue.put((min_id, max_id))
+
+    # Effectively a sentinel value to let chunk processors know they've reached the end of the queue.
+    await queue.put(END_OF_QUEUE)
+
+    return queue
+
+
+async def process_queue(queue, original_queue_size, processed_queue, pool):
+    while True:
+
+        min_id, max_id = await queue.get()
+        if min_id is None:
+
+            # Need to add the sentinel value back to the queue so other pipelines can pick it up
+            # else they'll wait forever for nothing.
+            await queue.put(END_OF_QUEUE)
+            break
+
+        with Timer() as t:
+            sql = SQL.format(min_id=min_id, max_id=max_id)
+            async with pool.acquire() as connection:
+                result = await connection.execute(sql)
+
+        await processed_queue.put((min_id, max_id))
+
+        affected = int(result.split()[-1])
+
+        ratio = processed_queue.qsize() / original_queue_size
+        estimated_remaining_runtime = overall.as_string(overall.estimated_remaining_runtime(ratio))
+        message = "Updated {:,} transactions from {} through {} in {} - estimated remaining runtime is {}".format(
+            affected, min_id, max_id, t, estimated_remaining_runtime
+        )
+        logging.info(message)
+
+
+async def run_pipelines(queue, original_queue_size, processed_queue, loop, pool):
+    tasks = [
+        asyncio.ensure_future(process_queue(queue, original_queue_size, processed_queue, pool), loop=loop)
+        for _ in range(PIPELINES)
+    ]
+    await asyncio.gather(*tasks)
+
+
+async def main(loop):
+    async with asyncpg.create_pool(dsn=CONNECTION_STRING, loop=loop) as pool:
+        await update_sql_functions(pool)
+        min_id, max_id = await get_min_max_ids(pool)
+        to_process_queue = await create_chunk_queue(min_id, max_id, loop)
+        processed_queue = asyncio.Queue(loop=loop)
+        await run_pipelines(to_process_queue, to_process_queue.qsize() - 1, processed_queue, loop, pool)
+
+
+if __name__ == "__main__":
+    with Timer("Overall run") as overall:
+        _loop = asyncio.get_event_loop()
+        try:
+            _loop.run_until_complete(main(_loop))
+        finally:
+            _loop.close()
```

Comment:
@kbard This whole `asyncio` and `asyncpg` queueing setup is excellent! I'd live to see a deep-dive walkthrough of this in a backend dev meeting.